ops/main.py

```diff
@@ -37,9 +48,37 @@ def main(charm_class: Type[ops.charm.CharmBase], use_juju_for_storage: Optional[
 
     See `ops.main() <#ops-main-entry-point>`_ for details.
     """
-    warnings.warn(
-        'Calling `ops.main.main()` is deprecated, call `ops.main()` instead',
-        DeprecationWarning,
-        stacklevel=2,
+    # Normally, we would do warnings.warn() with a DeprecationWarning, but at
+    # this point in the charm execution, the framework has not been set up, so
+    # we haven't had a chance to direct warnings where we want them to go. That
+    # means that they'll end up going to stderr, and with actions that means
+    # they'll end up being displayed to the user.
+    # This means that we need to delay emitting the warning until the framework
+    # has been set up, so we wrap the charm and do it on instantiation. However,
+    # this means that a regular warning call won't provide the correct filename
+    # and line number - we could wrap this in a custom formatter, but it's
+    # simpler to just use a logging.warning() call instead.
+    # Note that unit tests almost never call main(), so wouldn't show this
+    # warning, so we're relying on integration tests, which aren't going to be
+    # treating warning calls in any special way. That means that we can just
+    # output the warning directly as a log message, and the effect will be the
+    # same.
+    # Note also that this will be logged with every event. Our assumption is
+    # that this will be noticeable enough during integration testing that it
+    # will get fixed before going into production.
+    frame = _top_frame()
+    assert frame is not None
+
+    class DeprecatedMainCharmBase(charm_class):
+        def __init__(self, *args: Any, **kwargs: Any):
+            super().__init__(*args, **kwargs)
+            logger.warning(
+                '%s:%s: DeprecationWarning: Calling `ops.main()` is deprecated, '
+                'call `ops.main()` instead',
+                frame.f_code.co_filename,
+                frame.f_lineno,
+            )
+
+    return _main.main(
+        charm_class=DeprecatedMainCharmBase, use_juju_for_storage=use_juju_for_storage
```

Comment:
If it's used in this way, the new class should have a different name.

Alternatively, we could declare a new charm class dynamically with name being upstream name + some hack, and the deprecated base (or mixin?) added in.

Reply:
The class did have a new name. I guess it could have been dynamic rather than fixed, but I'm not sure whether that would really have any impact. But gone now anyway.